### ISSUE TITLE
Fix check subdue (broken and pulled prior to GA)

### ISSUE THREAD

**Issue by echlebek** [sensu/sensu-go#1955]
## The fix

https://github.com/sensu/sensu-go/issues/1955#issuecomment-1074144897

## Context

After spending some time debugging a test that only failed during certain times of day, I was finally able to isolate a test that I think behaves reliably.

The test that fails is `TestIsSubdued/valid_subdue`.

The reproduction I've created, which should hopefully fail any time of day, is here:

```
func TestTimeWindowWhenEarlyMorning(t *testing.T) {
	now, err := time.Parse(time.RFC1123Z, "Mon, 13 Aug 2018 5:30:00 -0900")
	if err != nil {
		t.Fatal(err)
	}
	subdue := &TimeWindowWhen{
		Days: TimeWindowDays{
			All: []*TimeWindowTimeRange{
				{
					Begin: now.Add(-time.Hour).Format(time.Kitchen),
					End:   now.Add(time.Hour).Format(time.Kitchen),
				},
			},
		},
	}
	subdued, err := subdue.InWindows(now)
	if err != nil {
		t.Fatal(err)
	}
	require.True(t, subdued)
}
```

Because `Begin` is an hour before `now`, and `After` is an hour after, the subdue should be true.

I believe the problem is that we are not consistently handling times in UTC. For instance, Check.IsSubdued does not use UTC, whereas the unit test TestIsSubdued does (to set up the test). In the test that I pasted to this issue, I have not used UTC, and observed a failure.

The InWindow method assumes that the time it receives is in UTC, so I think that something is not quite right.

What I can't figure out is why the TestIsSubdued test does not fail every time - it should have a constant offset from UTC. Because it only fails at certain times of day, I am concerned that may be a deeper problem with the InWindow method. However, at this time, that is just conjecture.

Note that the TestIsSubdued test will never fail on CI systems, because they are on UTC. This really is a nasty bug!

**Comment by echlebek** [comment]
I'd like to try to summarize this issue from the user's point of view.

As a user, I set up sensu-backend on a server that happens to be running in the PST timezone.

I decide that I don't want to be woken up early in the morning, so I create a check subdue from 4:30am to 6:30 am like so:

```json
{
  "days": {
    "all": [
      {
        "begin": "4:30 AM",
        "end": "6:30 AM"
      }
    ]
  }
}
```

I apply this data to the check I want to subdue and go to bed. I get woken up at 5 AM, because the subdue didn't evaluate to true.

Confused, later on in the day, I test some other check subdues: they appear to work fine.

I reach out to Sensu support. They look at my issue, and after some confusion, recommend that I use UTC on my server.

I don't mind using UTC on my server, but I would like to think about times in my local timezone. I now have to mentally map times to UTC when create subdue definitions.

**Comment by echlebek** [comment]
I believe this issue is the result of several factors.

1) We do not have a way to unambiguously transmit datetimes to the sensu backend. Rather, we rely on strings like "4:30 AM", and then try to get it right with some somewhat complex business logic.

2) We assume that times are UTC in some places, but not in all places.

3) We make decisions about the scope of time windows based on the current time. I believe this could be one of the reasons for the inconsistent behaviour. (See https://github.com/sensu/sensu-go/blob/master/types/time_window.go#L87)

**Comment by echlebek** [comment]
I'd like to propose a solution to this issue, that changes how users specify subdue.

Instead of using the current TimeWindowWhen format (similar to above), I wonder if we could implement a simpler system that is easier to verify, by having a similar UX to calendaring products such as Google Calendar.

1) A user creates a single time window at a specific date and time. Say, August 14th, 2018 4:00 AM to August 14th, 2018 6:00 AM.

2) The user is presented with the option of repeating the time window. The options provided will be Daily, Weekly, Monthly, Yearly, or a selection of days of the week.

The client software will compute the duration of the intervals selected.

The data transmitted to the server will then be unambiguous. For example, here is what would be transmitted to subdue a check from Monday to Friday, from 4:00 am to 6:00 am.

```json
{
  "start": "2018-08-14T4:00:00-08:00:00",
  "end": "2018-08-14T6:00:00-08:00:00",
  "repeat_calendar_days": [
    "monday",
    "tuesday",
    "wednesday",
    "thursday",
    "friday"
  ]
}
```

And here is what would be transmitted to repeat the time window every week:

```json
{
  "start": "2018-08-14T4:00:00-08:00:00",
  "end": "2018-08-14T6:00:00-08:00:00",
  "repeat_weekly": true
}
```

If we send unambiguous datetimes to the server, then it doesn't matter if the server is using UTC or another time zone - the Go time library will automatically handle the offset for us. For example: https://play.golang.org/p/xDClN-oKjWx

This is a breaking change. If we decide to move ahead with this, we should probably do it before GA. Otherwise, making this change will present additional difficulties.

This proposal also dovetails with the idea of allowing users to schedule events in Sensu with external calendaring software.

**Comment by jamesdphillips** [comment]
This seems inline with (what I imagine) my expectations would be as an end-user.

One suggestion: we should also consider adding a `location` field to the time window, that way it can adjust to the author's expectations of day-light savings and standard time (if applicable.)

**Comment by echlebek** [comment]
The ISO-8601 (also see RFC-3339) format encodes the timezone as a numeric offset already, which resolves a lot of issues with named timezones, so we actually don't need a location field. Non-numeric locations are actually quite troublesome to deal with.

This might present an issue with users who expect automatic daylight savings time adjustments - we should consider taking that into consideration, so thanks for pointing that out!

**Comment by jamesdphillips** [comment]
> The ISO-8601 (also see RFC-3339) format encodes the timezone as a numeric offset already, which resolves a lot of issues with named timezones, so we actually don't need a location field.

My apologies I may not have been clear, I don't see any issue with accepting ISO-8601 compatible dates. But, unless I am missing something obvious in my mental model.. local offsets (-800) are only relevant for the moment in time they are referencing and do no provide us with enough information to adjust in the future. With that in mind, I would argue we require a location field. Eg.

```json
{
  "start": "2018-08-14T4:00:00-08:00:00",
  "end": "2018-08-14T6:00:00-08:00:00",
  "repeat_calendar_days": [
    "monday",
    "tuesday",
    "sunday",
  ],
  "tz | location | tzinfo | idk": "Americas/Vancouver" // optional
}
```

> Non-numeric locations are actually quite troublesome to deal with.

Without justification present I'm unconvinced, at least on unix platforms it looks like golang simply uses the tz database that (from my understanding) should be present on all distributions.

**Comment by jamesdphillips** [comment]
..mind you.. [this fallback](https://github.com/golang/go/blob/master/src/time/zoneinfo_unix.go#L28-L34) may complicate things in a clustered environment.

**Comment by echlebek** [comment]
@jamesdphillips it gets complicated when someone wants to use a common timezone abbreviation that actually maps to more than one timezone. Or a common timezone name that doesn't actually exist in the tzdata - such as PST, which is actually represented as PST8PDT (deprecated) in tzdata. Otherwise you're right - the standard location names work fine.

If we gave people a list to select from, that would address all of my concerns about a location field.

**Comment by jamesdphillips** [comment]
For what it's worth, my concern with the previous proposal was my anticipation of receiving an angry bug report along the lines of: "I was awoken an hour early because of the timezone change". After looking at this issue again, while I do still feel my concern is somewhat valid, I think we could easily add timezone support in later, without it being a breaking change.

The Google Calendar API provides a good example; require a TZ offset _unless_ a valid time zone from the IANA db is provided. 

<img width="600" alt="Google Calendar API - start.timeZone" src="https://user-images.githubusercontent.com/194892/81018873-410ebc00-8e1a-11ea-8a56-15cb444e5485.png">

**Comment by portertech** [comment]
![backfromdead](https://user-images.githubusercontent.com/149630/158440434-1afd9c29-6c1e-4795-96f6-bdfe97a5b267.gif)

**Comment by echlebek** [comment]
![whatyear](https://user-images.githubusercontent.com/1633772/158440774-d9c399a6-6f12-47bb-9bc8-2e0dbc8520e9.gif)

**Comment by echlebek** [comment]
![coffinrise](https://user-images.githubusercontent.com/1633772/158441152-50872661-b729-418e-b6b9-a470aa8e6652.gif)

**Comment by calebhailey** [comment]
Better late than never? 😅 I think I'm going to owe some of our Sensu friends some 🍻 for letting this one linger for so long. 

I like the proposal to require TZ specificity in subdue schedules. 

Some of the earlier comments in this thread have got me wondering if we ought to further simplify the configuration by removing the concept of an "end time" and adopt a "duration" instead? Less chance of users creating a subdue (scheduling exception) that ends before it starts? 

I also think we should reduce the scope of what we support in terms of "repeat" configuration. It should be limited to a fixed set of options, perhaps as follows: 

- `annually` (same date, every year)
-  `monthly` (same day of month, every month)
- `weekly` (same day of week, every week)
- `weekdays` (same time of day, M-F)
- `weekends` (same time of day, Sat-Sun)
- `daily` (same time of day, every day)
- `mondays`, `tuesdays`, `wednesdays`, `thursdays`, `fridays`, `saturdays`, `sundays` (same time of day, on the specified days)

Example: 

```
[
  {
    "begin": "2022-03-17T11:00:00-0700",
    "duration": "3600",
    "repeat": [
      "weekly"
    ]
  }
]
```

What's the difference between this example ("weekly") and the specific day of week? With the latter users will be able to configure weekly schedule on multiple days of the week; for example: 

```
[
  {
    "begin": "2022-03-17T11:00:00-0700",
    "duration": "3600",
    "repeat": [
      "thursdays",
      "fridays"
    ]
  }
]
```

**Comment by calebhailey** [comment]
After further consideration, I would suggest we proceed as suggested by @echlebek and require timestamps with UTC offset for `begin` and `end`, with a simplified "repeat" scheduling implementation. 

Example subdue schedule: 

```yaml
subdue:
  - begin: 2022-03-21T16:04:00-0700
    end: 2022-03-21T18:58:00-0700
    repeat:
      - mondays
      - tuesdays
      - wednesdays
      - thursdays
  - begin: 2022-03-25T16:04:00-0700
    end: 2022-03-25T23:59:59-0700
    repeat:
      - fridays
  - begin: 2022-03-26T00:00:00-0700
    end: 2022-03-26T23:59:59-0700
    repeat:
      - saturdays
  - begin: 2022-03-27T00:00:00-0700
    end: 2022-03-27T16:58:00-0700
    repeat:
      - sundays
```

When a repeat schedule is provided, the `begin` timestamp should only be used for the time of day, and the delta between `begin` and `end` is used to determine the duration of the subdue. 

Examples:

* The following example applies to Sunday, March 20st (PT).

  ```yaml
  subdue:
    - begin: 2022-03-20T16:04:00-0700
      end: 2022-03-20T18:58:00-0700
  ```

* The following example — with the same `begin` timestamp as the previous example – is applied every week from Monday through Thursday after Sunday March 20st (PT), but **does not** apply to March 20th.

  ```yaml
  subdue:
    - begin: 2022-03-20T16:04:00-0700
      end: 2022-03-20T18:58:00-0700
      repeat:
        - mondays
        - tuesdays
        - wednesdays
        - thursdays
  ```

**Comment by portertech** [comment]
Initial implementation:

- Update check type and validation with the modified subdue spec
- Update check scheduler subdue implementation to use the new subdue spec

**Comment by portertech** [comment]
Closed by #4680
